pep-0635.rst

```diff
@@ -0,0 +1,689 @@
+OO style -- unlike Java, Python encourages defining functions at the
+top-level of a module, and for simple data structures, tuples (or
+named tuples or lists) and dictionaries are often used exclusively or
+mixed with classes or data classes.
+
+Pattern matching is particularly suitable for picking apart such data
+structures.  As an extreme example, it's easy to write code that picks
+a JSON data structure using ``match``.
+
+TODO: Example code.
+
+
+
+Motivation (Tobias' version)
+============================
+
+**This section should explain why we think pattern matching is a good
+addition for Python.**
+
+Since Python is a dynamically typed language, Python code frequently
+has to deal with data that comes in different forms and shapes.  This,
+in turn, gives rise to a high degree of versatility by favouring a
+'duck typing' style, imposing only the bare minimum of requirements
+on the form and shape of data, i.e. the structure of data objects.
+Nonetheless, we find that Python code is often sparkled with
+conditions depending on ``isinstance``, ``len``, ``getattr``,
+``hasattr``, etc.  Despite the benefits of 'duck typing', actual
+code still requires to query the format or type of an object in order
+to choose an appropriate processing action or to extract relevant bits
+of information.
+
+Unfortunately, the conditions for slightly more complex structures
+quickly grow into sequences of partly interdependent bits of
+structural tests, hurting readbility and maintability.  Pattern
+matching offers here a more direct scheme of expressing the minimal
+structure that an objects needs to have in order to allow for further
+processing.  Rather than writing a series of manual tests, patterns
+follow a *declarative* style, which is well-known for improved
+readability, maintability, and for delegating the burden of efficient
+execution on the compiler and interpreter.
+
+The concept of pattern matching is similar to regular expressions,
+where succinct patterns describe a textual structure.  A dedicated
+compiler then transforms these declarative patterns into highly
+efficient finite state machines.  In contrast to regular expressions,
+pattern matching targets Python objects rather than textual data,
+and builds on *decision trees* as the driving motor for finding a
+match.  Moreover, pattern matching blends the matching of a suitable
+pattern with that of a function, i.e. code that is executed in order
+to handle and process the information provided by a specific kind of
+object.
+
+One of the simplest forms of pattern matching that we find in other
+languages comes in the form of *function overloading*.  The type and
+number of arguments then determine which implementation of a specific
+function will be executed.  Object-oriented languages (including
+Python) may also use the *visitor pattern* to differentiate an action
+based on the type or class of an object.  Both of these approaches,
```

Comment:
I'm not sure I want to be calling out the visitor pattern here, since it is one of the more complicated OO patterns (see https://en.wikipedia.org/wiki/Visitor_pattern#Python_example).

A simpler analogy is just method overloading, which also differentiates based on the class. And if you want to have the code outside the classes, a pattern people often recommend (though I find it ugly too) is a dict full of lambdas (or other functions) indexed by class.

Visitor pattern, method overloading and dict-of-lambdas all have in common that they only switch on a single type, and any logic to switch on multiple types or on a type and some other property (e.g. len()) must be put inside the method/lambda. So that's our key argument.

(Function overloading is as powerful as a sequence pattern filled with argument-less class patterns. That's better, but still not enough.)

Reply:
Yes, I am aware of the visitor pattern on the Wikipedia page and find it really ugly and "_unpythonic_" (I am very caucious with this term, but what I mean here is that the code there goes out of its way to copy and imitate Java as closely as possible, without taking the strengths and nature of Python into consideration at all; similar to the "you can write assembly code in any language" ^_^).

With "method overloading", are you referring to polymorphism or Python's answer to the visitor pattern (how it is done in the AST module)?

Anyway, since there has been some criticism on the mailing list that the visitor pattern already works well and pattern matching is not needed, I feel we should address it somehow and contrast the two (and possible other approaches).

> Visitor pattern, method overloading and dict-of-lambdas all have in common that they only switch on a single type [...]. So that's our key argument.

Yes, I fully agree.

Reply:
Yes, the visitor pattern is very ugly -- we're using it extensively in mypy and while I agree it's probably the best approach for mypy's needs, it is hard to get it right, and there's a lot of boilerplate involved, making it indeed very unpythonic.

**But...** I think this is too early in the PEP to talk about the visitor pattern (in part because it is so ugly that it feels like a strawman). Yes, we should address it, but preferably quite a bit later in the PEP.

By "method overloading" I just meant adding methods to the subject class. I think the word I was using was "overriding" not "overloading", sorry. This has problems that the visitor pattern is trying to address (by moving the application-specific code out of the central class hierarchy), but nevertheless it is commonly used (e.g. when separating app from library is not deemed important) and much more well-known than the visitor pattern. Dict-full-of-lambdas (or functions) is probably more well-known among Python developers, since it's the "FAQ" answer to "how do I do a switch" (using a class as the key is just one variant of that).

Anyway, your point seems to be that OO languages allow switching on the class of an object using the visitor pattern -- to which I say "duh, you can do that by adding methods". So if you want to contrast patterns with something that can only switch on the class, the simplest example seems to be plain methods. If someone says "but the visitor pattern is more powerful" (which it is, though at terrible cost) our answer is just going to be "but it still only switches on the class".

Another thing that worries me (still) about competing with the visitor pattern is https://github.com/gvanrossum/patma/issues/137. As you may recall, I worked out an example that used to use the visitor pattern and it came down to having a match with N cases where N is the number of classes in the central class hierarchy. Certainly our current implementation makes N isinstance() calls, whereas method dispatch uses log(N) dict lookups.

I recall (though I can't find the code anywhere) doing a larger example (N=60?) and finding that the match/case code was faster than the original code using the visitor pattern. I think it was because the visitor pattern doubles the number of calls (for every visit() call there's an accept() call), but I still can't help worrying about match being O(N) vs. visitor's O(log(N)).

(And sorry for the rant.)

Reply:
I think we are essentially on the same page here.

Although the visitor pattern is ugly, it does have its places.  I think the different between method overriding/polymorphism and the visitor pattern is that the visitor pattern is a means to add quasi-methods to a class without having to actually touch the class.  From my point of view, it is not so much about one being simpler than the other, but more a question of grouping inside the class hierarchy (i.e. the visitor pattern lets you group your code by method rather than by class).

> but it still only switches on the class

Fully agreed.  And just in case it should come up: even if you incorporate more than just the class into the switch mechanism, you have to write quite some boilerplate and once it is there, everthing has to follow the exact same structure.  In short: pattern matching is great because it is flexible, can work with highly inhomogeneous and non-uniform cases, and has the compiler take over the boilerplate stuff.

> but I still can't help worrying about match being O(N) vs. visitor's O(log(N))

Thinking about that, I feel it is one of those apple-oranges things.  And here is why: the visitor pattern is a specific "algorithm" for solving a problem if you will.  We have been calling out other means such as method overriding, if-elif-chains, and lambda-dictionaries that could achieve the same goal.  Pattern matching, on the other hand, is a matter of "declaration".  It does not really specify how the interpreter picks the correct case clause.  So, in principle, a compiler could choose to implement pattern matching through a visitor pattern under the hood (isn't that the prevalent way of emulating overloaded functions in Python?).  Anyway, I would argue that pattern matching as such does not have a specific runtime complexity.

Reply:
>  Pattern matching, on the other hand, is a matter of "declaration". It does not really specify how the interpreter picks the correct case clause.

All I know is that our current implementation is O(N), and doing better would require a lot of work. Probably an upper bound is O(D) where D is the inheritance depth of the typical target class (I have an example where it's necessary to walk the entire MRO).

> So, in principle, a compiler could choose to implement pattern matching through a visitor pattern under the hood (isn't that the prevalent way of emulating overloaded functions in Python?).

I don't know how people typically implement overloaded functions; I once wrote a simple multidispatch that used a cache dict whose keys are tuples of classes, and on a cache miss I'd do a linear scan over all overloads. Assuming a fixed number of classes and a long computation that could be O(1), and it certainly feels simpler than building a visitor pattern under the hood.

> Anyway, I would argue that pattern matching as such does not have a specific runtime complexity.

But our reference implementation does, and that's likely what will stick around for a long time.